cime/src/drivers/mct/main/cime_comp_mod.F90

```diff
@@ -2523,8 +2546,94 @@ subroutine cime_run()
        !----------------------------------------------------------
        !| GLC SETUP-SEND
        !----------------------------------------------------------
+       if (glc_present) then
+
+          if (ocn_c2_glc) then
+             call prep_glc_calc_o2x_gx(timer='CPL:glcprep_ocn2glc') !remap ocean fields to o2x_g at atmospheric timestep
+          end if
+          if (lnd_c2_glc) then
+             ! Note that l2x_gx is obtained from mapping the module variable l2gacc_lx
+             call prep_glc_calc_l2x_gx(fractions_lx, timer='CPL:glcprep_lnd2glc')
+                !...and then here is where Bill S. does the new in-coupler downscaling.  I changed this so it's downscaling
```

Comment:
@billsacks , do you have any concerns with the change described in this comment?

Reply:
Yes, I do have concerns about this, but I haven't fully digested these changes. This was what motivated me to ask you to reconcile this with the code in `cime_run_glc_setup_send`: it's hard for me to see what changes were intentional and what were unintentional. What I'd suggest is that you and @JeremyFyke reconcile these routines in the way that you'd like them, then I'll review that reconciliation and comment on anything that I think will cause trouble. Does that sound okay to you?

Reply:
@billsacks , specifically, what is going on here is that before you had the following order of operations:
1. `prep_glc_accum` (via `cime_run_lnd_recv_post`)  (accumulate fields on LND grid)
2. `prep_glc_accum_avg` (average fields on LND grid)
3. `prep_glc_calc_l2x_gx` (remap to GLC grid)
4. `prep_glc_mrg` (finalize remapped LND fields for usage by GLC)

@JeremyFyke 's changes reorder operations like this:
1. `prep_glc_calc_l2x_gx` (remap to GLC grid)
2. `prep_glc_mrg` (finalize remapped LND fields for usage by GLC)
3. `prep_glc_accum`  (accumulate fields)
4. `prep_glc_accum_avg` (average fields)

Looking at it closer, it looks like steps 3 and 4 in the new ordering are still happening on the LND grid, which would be a problem.

I believe the goal of the reordering was so there would be common `accum` and `accum_avg` steps for fields coming from both LND and OCN.  I think to do this, we'd have to modify those routines to work on the GLC grid for the fields coming from LND instead of on the LND grid.  Alternatively, we could keep the old order of operations for the LND coupling fields and then  have different versions of the  `accum` and `accum_avg` routines for the fluxes coming from LND vs. OCN. 

Let me know if you have any thoughts on this.

Reply:
@billsacks , my questions were motivated by my attempts to reconcile them.  It turns out most of `cime_run_glc_setup_send` gets gutted to reconcile the changes due to the order of operations change described in my previous comment.  (It's not possible to just reorganize inside `cime_run_glc_setup_send` because some of the steps have to move outside the if statement enclosing the call to `cime_run_glc_setup_send`.)   If my previous comment about the order of operations is not specific enough, I don't mind attempting to reconcile `cime_run_glc_setup_send`, but it might not look much simpler than it already does.

Reply:
I don't completely understand all of the new code / changes, but: At least in CESM, we want to accumulate the lnd2glc fields on the land grid, and then only do the regridding once per year - i.e., based on `glcrun_avg_alarm`. This is important for both performance and scientific reasons (the latter because the lnd -> glc regridding is not a linear operation, because of the conservation corrections that are done in this remapping). So, I'm not sure how the new code needs to be organized to achieve that, but that's an important requirement to maintain.

It seems like you need separate prep_glc_accum routines to (1) accumulate fields from lnd2glc (done after lnd_run), and (2) accumulate fields from ocn2glc (done after ocn_run). So you could have something like `prep_glc_accum_from_lnd` and `prep_glc_accum_from_ocn`. It's probably still fine to have a single `prep_glc_accum_avg` routine, because that should only be needed right before running glc.

Does that help at all?

Regardless of what's done here, it would still be good to maintain the new cime_run design of having most work done by subroutines rather than having big blocks of code inlined in there - even if this requires adding one or two new subroutines holding the new code that has been added in cime_run.

Reply:
@billsacks , thanks for the input on this - that is actually very helpful.  Knowing that maintaining the existing functionality of accumulating lnd2glc fields on the land grid is a requirement makes it clear that the accumulation changes currently in the PR will not work.  Thanks for pointing out the scientific reason why that's not ok - that makes sense to me.  

I've been coming to the same conclusion that separate prep_glc_accum routines will be needed to handle this situation.  I was looking around for an analogous situation to what we have here elsewhere in the model, and I don't think there is one.  Namely, a component receiving fluxes from two other components that have different coupling intervals from one another.  As far as I can tell no other component has this situation where accumulation has to happen differently for different components from which it receives information.

I'm beginning to have a sense of how this area of the code needs to be refactored.  I'll try to draft up and test changes, and then I'll ask you to look over the proposal.  Might take a few days.

Reply:
Thanks a lot, @matthewhoffman - this sounds good.

I'm thinking that the ocean receives and accumulates fields from components with different coupling frequencies (atm vs. rof), and in that case it looks like it accumulates fields on the ocn grid. But yeah, with the custom remapping we do from lnd -> glc, I don't think that'll work, and we need something different.

For what it's worth, I ran this by Mariana:

> It seems like you need separate prep_glc_accum routines to (1) accumulate fields from lnd2glc (done after lnd_run), and (2) accumulate fields from ocn2glc (done after ocn_run). So you could have something like prep_glc_accum_from_lnd and prep_glc_accum_from_ocn. 

and she agreed that that sounds like the right way to go.

Reply:
@billsacks , I've restructured the code in cime_comp_mod.F90 to deal with issues we've been discussing.  I've tested to make sure it compiles and the subshelf melt flux calculation still works, but I have not yet done more extensive testing.  I will work on your other code cleanup requests next and then do more thorough testing.  But I wanted to point out the two new commits I added now, in case you want to quickly look them over and make sure they address your major concerns.